pennylane/transforms/decompose.py

```diff
@@ -0,0 +1,157 @@
+    decomposer: Callable[[qml.operation.Operator], Sequence[qml.operation.Operator]],
+    max_expansion: Optional[int] = None,
+    current_depth=0,
+) -> Generator[qml.operation.Operator, None, None]:
+    """A generator that yields the next operation that is accepted."""
+    max_depth_reached = False
+    if max_expansion is not None and max_expansion <= current_depth:
+        max_depth_reached = True
+    if acceptance_function(op) or max_depth_reached:
+        yield op
+    else:
+        try:
+            decomp = decomposer(op)
+            current_depth += 1
+        except qml.operation.DecompositionUndefinedError as e:
+            raise UserWarning(
+                f"Operator {op.name} has no supported decomposition and is not in the gate set."
+            ) from e
+
+        for sub_op in decomp:
+            yield from _operator_decomposition_gen(
+                sub_op,
+                acceptance_function,
+                decomposer=decomposer,
+                max_expansion=max_expansion,
+                current_depth=current_depth,
+            )
+
+
+@transform
+def apply_decomposition(tape, gate_set=None, gate_rules=None, max_expansion=None):
+    """Decomposes quantum circuit into a desired gate set.
+
+    Args:
+        tape (QuantumScript or QNode or Callable): a quantum circuit.
+        gate_set (set, optional): A set of decomposition gates. Defaults to None. If ``None``, gate set defaults to all operators not currently present.
+        gate_rules (Callable[qml.operation.Operator, bool], optional): A rule set for the gate set to follow. Defaults to None.
+        max_expansion (int, optional): The maximum depth of the expansion. Defaults to None.
+
+    Returns:
+        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumScript], function]:
+
+        The decomposed circuit. The output type is explained in :func:`qml.transform <pennylane.transform>`.
+
+    .. seealso:: :func:`~.pennylane.devices.preprocess.decompose` for a transform that is designed particularly for restricted gate sets on specific device architectures.
+
+    **Examples:**
+
+    >>> tape = qml.tape.QuantumScript([qml.IsingXX(1.2, wires=(0,1))], [qml.expval(qml.Z(0))])
+    >>> batch, fn = apply_decomposition(tape, gate_set = {"CNOT", "RX", "RZ"})
+    >>> batch[0].circuit
+    [CNOT(wires=[0, 1]),
+    RX(1.2, wires=[0]),
+    CNOT(wires=[0, 1]),
+    expval(Z(0))]
+
+    >>> @partial(apply_decomposition, gate_rules = lambda obj: len(obj.wires) <= 2)
+    >>> @qml.qnode(device)
+    >>> def circuit():
+    >>>     qml.Toffoli(wires = range(3))
+    >>>
+    >>> print(qml.draw(circuit)())
+    0: ───────────╭●───────────╭●────╭●──T──╭●─┤
+    1: ────╭●─────│─────╭●─────│───T─╰X──T†─╰X─┤
+    2: ──H─╰X──T†─╰X──T─╰X──T†─╰X──T──H────────┤
+
+    """
+    universal_ops = set(qml.ops.__all__)
+
+    if gate_set is None:
+        gate_set = universal_ops - set(op.name for op in tape.operations)
```

Comment:
I chose to do this as if the user provides both `gate_set = None` and `gate_rules = None` and `max_expansion=1` for example, it will still decompose the circuit by one "stage". Although, this probably should be looked at further.

If the gate_set remained universal, then the decomposition would do nothing since all of the operations in the provided tape would be valid.